### PR TITLE
Added null check before NavigationRail.onDestinationSelected is called

### PR DESCRIPTION
**Pull request by wilsonowilson**
At the moment, NavigationRail.onDestinationSelected which is nullable is called regardless of whether it is null or not.

```dart
onTap: () {
    widget.onDestinationSelected!(i);
},
```
If it is null, the error "Null check operator used on a null value" is thrown.

This PR fixes that by adding a null check before it is called:

```dart
if(widget.onDestinationSelected != null)
    widget.onDestinationSelected!(i);
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/navigation_rail.dart [hunk 1]
```diff
@@ -498,7 +498,8 @@ class _NavigationRailState extends State<NavigationRail> with TickerProviderStat
                           labelTextStyle: widget.selectedIndex == i ? selectedLabelTextStyle : unselectedLabelTextStyle,
                           padding: widget.destinations[i].padding,
                           onTap: () {
-                            widget.onDestinationSelected!(i);
+                            if(widget.onDestinationSelected != null)
+                              widget.onDestinationSelected!(i);
```
Comment (creativecreatorormaybenot) [inline packages/flutter/lib/src/material/navigation_rail.dart:502]: Just a style choice, however, I feel this is more concise:

```suggestion
                            widget.onDestinationSelected?.call(i);
```

Comment (wilsonowilson) [reply packages/flutter/lib/src/material/navigation_rail.dart:502]: I was going to do that, but the BottomNavigationBar uses that style which made me opt in
```dart
 onTap: () {
    if (widget.onTap != null)
      widget.onTap!(i);
},
```

Comment (creativecreatorormaybenot) [inline packages/flutter/lib/src/material/navigation_rail.dart:502]: What about this then? :)

```suggestion
                            if (widget.onDestinationSelected != null)
                              widget.onDestinationSelected!(i);
```

Would be interesting to get insights on the style choice from a team member. Especially the null-assertion (bang) after the null check seems off.

**Review by Piinks — COMMENTED** [review]
Hey @wilsonowilson thanks for the contribution! Can you add a test that ensures we do not regress this later?

**Comment by wilsonowilson** [comment]
Sure! I just added a test

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 2]
```diff
@@ -1917,6 +1917,21 @@ void main() {
     await tester.tap(find.text('Ghi'));
     expect(selectedIndex, 2);
   });
+  testWidgets('onDestinationSelected is not called if null', (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/material/navigation_rail_test.dart:1920]:
```suggestion

  testWidgets('onDestinationSelected is not called if null', (WidgetTester tester) async {
```

File: packages/flutter/lib/src/material/navigation_rail.dart [hunk 3]
```diff
@@ -498,7 +498,8 @@ class _NavigationRailState extends State<NavigationRail> with TickerProviderStat
                           labelTextStyle: widget.selectedIndex == i ? selectedLabelTextStyle : unselectedLabelTextStyle,
                           padding: widget.destinations[i].padding,
                           onTap: () {
-                            widget.onDestinationSelected!(i);
+                            if(widget.onDestinationSelected != null)
```
Comment (Piinks) [inline packages/flutter/lib/src/material/navigation_rail.dart:501]:
```suggestion
                            if (widget.onDestinationSelected != null)
```

**Review by Piinks — COMMENTED** [review]
Thanks for adding a test! It looks like the analyzer found some trailing white space. Also, is there an open issue about this that we can link to?

**Comment by wilsonowilson** [comment]
I couldn't find an open issue about this.

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 4]
```diff
@@ -1918,6 +1918,22 @@ void main() {
     expect(selectedIndex, 2);
   });
 
+  testWidgets('onDestinationSelected is not called if null', (WidgetTester tester) async {
+    const int selectedIndex = 0;
+
```
Comment (creativecreatorormaybenot) [inline packages/flutter/test/material/navigation_rail_test.dart:1923]:
```suggestion

```

Comment (creativecreatorormaybenot) [reply packages/flutter/test/material/navigation_rail_test.dart:1923]: (this is what the analyzer complains about)

Comment (wilsonowilson) [reply packages/flutter/test/material/navigation_rail_test.dart:1923]: Got it! Thanks!

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
Thank you for the contribution!

**Pull request merged**